curvefs/src/client/rpcclient/metaserver_client.cpp

```diff
@@ -819,6 +829,49 @@ void MetaServerClientImpl::UpdateXattrAsync(const Inode &inode,
     excutor->DoAsyncRPCTask(taskDone);
 }
 
+bool MetaServerClientImpl::ParseS3MetaStreamBuffer(butil::IOBuf* buffer,
+                                                   uint64_t* chunkIndex,
+                                                   S3ChunkInfoList* list) {
+    butil::IOBuf out;
+    std::string delim = ":";
+    if (buffer->cut_until(&out, delim) != 0) {
+        LOG(ERROR) << "invalid stream buffer: no delimiter";
+        return false;
+    } else if (!StringToUll(out.to_string(), chunkIndex)) {
+        LOG(ERROR) << "invalid stream buffer: invalid chunkIndex";
+        return false;
+    } else if (!list->ParseFromString(buffer->to_string())) {
+        LOG(ERROR) << "invalid stream buffer: invalid s3chunkinfo list";
+        return false;
+    }
+
+    return true;
+}
+
+bool MetaServerClientImpl::HandleS3MetaStreamBuffer(butil::IOBuf* buffer,
+                                                    S3ChunkInfoMap* out) {
+    uint64_t chunkIndex;
+    S3ChunkInfoList list;
+    if (!ParseS3MetaStreamBuffer(buffer, &chunkIndex, &list)) {
+        return false;
+    }
+
+    auto merge = [](const S3ChunkInfoList& from, S3ChunkInfoList* to) {
+        for (size_t i = 0; i < from.s3chunks_size(); i++) {
+            auto chunkinfo = to->add_s3chunks();
+            *chunkinfo = std::move(from.s3chunks(i));
```

Comment:
> this might not work, because the result of `from.s3chunks(i)` is const, and you can move from a const variable.

fixed.

Reply:
> this might not work, because the result of `from.s3chunks(i)` is const, and you can move from a const variable.

fixed:
```
 *chunkinfo = std::move(*from->mutable_s3chunks(i));
```